src/Build/Instance/TaskFactories/AssemblyTaskFactory.cs

```diff
@@ -393,10 +400,14 @@ public void CleanupTask(ITask task)
 #pragma warning restore SA1111, SA1009 // Closing parenthesis should be on line of last parameter
 
 #if FEATURE_APPDOMAIN
-                if (taskAppDomain != null)
+                if (taskAppDomain != null && taskInstance != null)
                 {
                     _tasksAndAppDomains[taskInstance] = taskAppDomain;
                 }
+                else if (taskAppDomain != null)
+                {
+                    AssemblyLoadsTracker.StopTracking(taskAppDomain);
```

Comment:
Do you still need _tasksAndAppDomains[taskInstance] = taskAppDomain; in this condition?

Reply:
No - it's just a cleanup if the `taskInstance` is `null` and `taskAppDomain` is not (in such a case we do not want to track as task was not created).
The condition is altered as there was a hidden bug in the original code (indexing with `null` key `taskInstance`)